src/testing/startup/db.rs

```diff
@@ -0,0 +1,108 @@
+use crate::db::{models::api_tokens::APITokenData, schema::api_tokens};
+use assert_fs::{fixture::PathChild, TempDir};
+use diesel::{connection::Connection, prelude::*};
+use std::io;
+use std::path::Path;
+use std::path::PathBuf;
+use thiserror::Error;
+
+pub struct DbBuilder {
+    migrations_folder: Option<PathBuf>,
+    token: Option<APITokenData>,
+}
+
+impl DbBuilder {
+    pub fn new() -> Self {
+        Self {
+            migrations_folder: None,
+            token: None,
+        }
+    }
+
+    pub fn with_token(&mut self, token: APITokenData) -> &mut Self {
+        self.token = Some(token);
+        self
+    }
+
+    pub fn with_migrations_from<P: AsRef<Path>>(&mut self, migrations_folder: P) -> &mut Self {
+        self.migrations_folder = Some(migrations_folder.as_ref().into());
+        self
+    }
+
+    fn create_db_if_not_exists(&self, db_path: &str) -> Result<(), DbBuilderError> {
+        rusqlite::Connection::open(db_path).map_err(DbBuilderError::CannotCreateDatabase)?;
+        Ok(())
+    }
+
+    fn do_migration(
+        &self,
+        connection: &SqliteConnection,
+        migration_folder: &PathBuf,
+    ) -> Result<(), DbBuilderError> {
+        let stdout = io::stdout();
+        let mut handle = stdout.lock();
+        diesel_migrations::run_pending_migrations_in_directory(
+            connection,
+            migration_folder,
+            &mut handle,
+        )
+        .map_err(DbBuilderError::MigrationsError)
+    }
+
+    fn try_do_migration(&self, connection: &SqliteConnection) -> Result<(), DbBuilderError> {
+        if let Some(migrations_folder) = &self.migrations_folder {
+            self.do_migration(&connection, migrations_folder)?;
+        }
+        Ok(())
+    }
+
+    fn try_insert_token(&self, connection: &SqliteConnection) -> Result<(), DbBuilderError> {
+        if let Some(token) = &self.token {
+            let values = (
+                api_tokens::dsl::token.eq(&(*token.token.as_ref())),
```

Comment:
I don’t understand this referencing/ de referencing dance. Nitpicking, but, can’t it be simplified?

Reply:
it was `cargo clippy` suggestion. I will take another look

Reply:
I think `api_tokens::dsl::token.eq(token.token.as_ref())` should be enough.